<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <meta http-equiv="X-UA-Compatible" content="ie=edge">
  <meta name="theme-color" content="#3367D6"/>
  <link rel="apple-touch-icon" href="/icons-192.png">
  <link rel="manifest" href="/manifest.json">
  
  <meta name="generator" content="Hexo 5.4.0">

  
    <meta name="description" content="生物信息学">
  

  

  
    <meta name="author" content="wenyuhao">
  

  

  

  <title>文玉昊</title>

  

  
    <link rel="shortcut icon" href="/favicon.ico">
  

  
    <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/typeface-source-code-pro@1.1.3/index.min.css">
    <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/highlightjs@9.16.2/styles/monokai.css">
  

  
<link rel="stylesheet" href="/css/style.css">

</head>
<body>
  <div class="root-container">
    
<!-- header container -->
<header class="header-container">
  
    <div class="post-image" style="background-image: url(/68686407_p0.jpg)"></div>
  

  <!-- navbar -->
<nav class="navbar">
  <div class="navbar-content">
    <!-- logo -->
    <div class="navbar-logo">
      <a href="/">
        
          文玉昊
        
      </a>
    </div>
    <!-- link -->
    <div class="navbar-link">
      <div class="navbar-btn">
        <div></div>
        <div></div>
        <div></div>
      </div>
      <ul class="navbar-list">
        
          <li class="navbar-list-item"><a href="/">首页</a></li>
        
          <li class="navbar-list-item"><a href="/about">关于</a></li>
        
      </ul>
    </div>
  </div>
</nav>

  
  

  
  

  
  

  
  

  
  

  
  
    <div class="header-content">
      <div class="post-text layout-block layout-padding">
        <h1 class="title-wrap"></h1>
        
      </div>
    </div>
  
  
</header>

    <!-- 分页(其他自定义页面) -->

<!-- 自定义页面 -->

<div class="body-container">
  <article class="content-container layout-block post-container">
    <div class="article-info">
      
      <section class="article-entry markdown-body layout-margin content-padding--large soft-size--large soft-style--box">
        <figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">import</span> pandas <span class="keyword">as</span> pd</span><br><span class="line">hg38=pd.read_table(<span class="string">&#x27;/data/wenyuhao/downloadTCGA/Homo_sapiens.GRCh38.94.gtf&#x27;</span>,skiprows=<span class="number">5</span>,header=<span class="literal">None</span>)</span><br><span class="line"><span class="function"><span class="keyword">def</span> <span class="title">deal</span>(<span class="params">xx</span>):</span></span><br><span class="line">    xx=xx.replace(<span class="string">&#x27; &#x27;</span>,<span class="string">&#x27;:&#x27;</span>).replace(<span class="string">&#x27;;:&#x27;</span>,<span class="string">&#x27;;&#x27;</span>).replace(<span class="string">&#x27;&quot;&#x27;</span>,<span class="string">&#x27;&#x27;</span>).strip(<span class="string">&#x27;;&#x27;</span>)</span><br><span class="line">    <span class="keyword">return</span> &#123;i.split(<span class="string">&#x27;:&#x27;</span>)[<span class="number">0</span>]:i.split(<span class="string">&#x27;:&#x27;</span>)[<span class="number">1</span>] <span class="keyword">for</span> i <span class="keyword">in</span> xx.split(<span class="string">&#x27;;&#x27;</span>)&#125;</span><br><span class="line">hg38[<span class="number">8</span>]=hg38[<span class="number">8</span>].apply(deal)</span><br><span class="line">jiuyt=pd.DataFrame(hg38[<span class="number">8</span>].to_list())</span><br><span class="line">sha=pd.concat([hg38,jiuyt],axis=<span class="number">1</span>)</span><br><span class="line">sha[:<span class="number">4</span>]</span><br><span class="line">sha1=sha[[<span class="number">0</span>,<span class="number">2</span>,<span class="number">3</span>,<span class="number">4</span>,<span class="string">&#x27;gene_name&#x27;</span>,<span class="string">&#x27;gene_id&#x27;</span>,<span class="string">&#x27;gene_biotype&#x27;</span>,<span class="string">&#x27;gene_version&#x27;</span>,<span class="string">&#x27;transcript_id&#x27;</span>,<span class="string">&#x27;transcript_name&#x27;</span>,</span><br><span class="line">     <span class="string">&#x27;transcript_biotype&#x27;</span>,<span class="string">&#x27;protein_id&#x27;</span>,<span class="string">&#x27;protein_version&#x27;</span>,<span class="string">&#x27;exon_id&#x27;</span>,<span class="string">&#x27;exon_version&#x27;</span>,<span class="string">&#x27;ccds_id&#x27;</span>]]</span><br><span class="line">sha1=sha1[sha1[<span class="string">&#x27;chr&#x27;</span>].apply(<span class="keyword">lambda</span> x: <span class="built_in">all</span>([i <span class="keyword">not</span> <span class="keyword">in</span> <span class="built_in">str</span>(x) <span class="keyword">for</span> i <span class="keyword">in</span> [<span class="string">&#x27;GL&#x27;</span>,<span class="string">&#x27;KI&#x27;</span>,<span class="string">&#x27;MT&#x27;</span>]]))].reset_index(drop=<span class="literal">True</span>)</span><br><span class="line">sha1[<span class="string">&#x27;chr&#x27;</span>]=sha1[<span class="string">&#x27;chr&#x27;</span>].astype(<span class="built_in">str</span>)</span><br><span class="line"><span class="comment">#去掉GL KI MT打头的染色体</span></span><br><span class="line">sha1[:<span class="number">5</span>]</span><br><span class="line">sha1.query(<span class="string">&#x27;gene_name==&quot;OR4F5&quot;&#x27;</span>)<span class="comment">#编码蛋白的基因</span></span><br><span class="line">sha1.query(<span class="string">&#x27;gene_name==&quot;ALG1L9P&quot;&#x27;</span>)<span class="comment">#不编码蛋白的基因</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#获取uniprot的序列</span></span><br><span class="line"><span class="keyword">import</span> requests <span class="keyword">as</span> rq</span><br><span class="line">uniprotSeq=<span class="keyword">lambda</span> x:rq.get(<span class="string">f&#x27;http://www.uniprot.org/uniprot/<span class="subst">&#123;x&#125;</span>.fasta&#x27;</span>).content.decode().replace(<span class="string">&#x27;\n&#x27;</span>,<span class="string">&#x27;&amp;nbsp&#x27;</span>,<span class="number">1</span>).replace(<span class="string">&#x27;\n&#x27;</span>,<span class="string">&#x27;&#x27;</span>).split(<span class="string">&#x27;&amp;nbsp&#x27;</span>,<span class="number">1</span>)</span><br><span class="line"><span class="comment">#获取ENST信息</span></span><br><span class="line"><span class="keyword">from</span> Bio <span class="keyword">import</span> SeqIO</span><br><span class="line"><span class="function"><span class="keyword">def</span> <span class="title">getENSTSeq</span>(<span class="params">enst</span>):</span></span><br><span class="line">    <span class="keyword">for</span> seq_record <span class="keyword">in</span> SeqIO.parse(<span class="string">&quot;/data/wenyuhao/55/genome/hg19GRCh37/gencode.v19.pc_translations.fa&quot;</span>, <span class="string">&quot;fasta&quot;</span>):</span><br><span class="line">        <span class="keyword">if</span>(<span class="built_in">str</span>(seq_record.<span class="built_in">id</span>).split(<span class="string">&#x27;|&#x27;</span>)[<span class="number">0</span>].split(<span class="string">&#x27;.&#x27;</span>)[<span class="number">0</span>]==enst):</span><br><span class="line">            <span class="keyword">return</span> <span class="built_in">str</span>(seq_record.seq)</span><br><span class="line">    <span class="keyword">return</span> <span class="string">&#x27;&#x27;</span></span><br></pre></td></tr></table></figure>


<p>为什么一个Gene symbol会对应多个ENSG?<br>在我的理解中Gene symbol和ENSG不是都在说基因吗，那么按道理是一对一的关系，但是在biomart ensembl中转，一个symbol对应了好几个ENSG。<br>比如说，GLIS2这个基因。</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">$ grep -n <span class="string">&#x27;GLIS2&#x27;</span> /data/wenyuhao/downloadTCGA/Homo_sapiens.GRCh38<span class="number">.94</span>.gtf | awk <span class="string">&#x27;&#123;print $10 $3&#125;&#x27;</span>  | grep gene</span><br><span class="line"><span class="string">&quot;ENSG00000126603&quot;</span>;gene</span><br><span class="line"><span class="string">&quot;ENSG00000262686&quot;</span>;gene</span><br></pre></td></tr></table></figure>
<p>但是去NCBI官网，发现只有一个ENSG00000126603，去GTF文件中寻找ENSG00000262686这个ID。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">$ grep -n &#x27;GLIS2&#x27; /data/wenyuhao/downloadTCGA/Homo_sapiens.GRCh38.94.gtf  | grep gene | tail -6</span><br><span class="line">2084322:16      havana  gene    4324667 4328340 .       -       .       gene_id &quot;ENSG00000262686&quot;; gene_version &quot;1&quot;; gene_name &quot;GLIS2-AS1&quot;; gene_source &quot;havana&quot;; gene_biotype &quot;lincRNA&quot;;</span><br><span class="line">2084323:16      havana  transcript      4324667 4328340 .       -       .       gene_id &quot;ENSG00000262686&quot;; gene_version &quot;1&quot;; transcript_id &quot;ENST00000576080&quot;; transcript_version &quot;1&quot;; gene_name &quot;GLIS2-AS1&quot;; gene_source &quot;havana&quot;; gene_biotype &quot;lincRNA&quot;; transcript_name &quot;GLIS2-AS1-201&quot;; transcript_source &quot;havana&quot;; transcript_biotype &quot;lincRNA&quot;; tag &quot;basic&quot;; transcript_support_level &quot;2&quot;;</span><br><span class="line">2084324:16      havana  exon    4328260 4328340 .       -       .       gene_id &quot;ENSG00000262686&quot;; gene_version &quot;1&quot;; transcript_id &quot;ENST00000576080&quot;; transcript_version &quot;1&quot;; exon_number &quot;1&quot;; gene_name &quot;GLIS2-AS1&quot;; gene_source &quot;havana&quot;; gene_biotype &quot;lincRNA&quot;; transcript_name &quot;GLIS2-AS1-201&quot;; transcript_source &quot;havana&quot;; transcript_biotype &quot;lincRNA&quot;; exon_id &quot;ENSE00002659645&quot;; exon_version &quot;1&quot;; tag &quot;basic&quot;; transcript_support_level &quot;2&quot;;</span><br><span class="line">2084325:16      havana  exon    4326430 4326494 .       -       .       gene_id &quot;ENSG00000262686&quot;; gene_version &quot;1&quot;; transcript_id &quot;ENST00000576080&quot;; transcript_version &quot;1&quot;; exon_number &quot;2&quot;; gene_name &quot;GLIS2-AS1&quot;; gene_source &quot;havana&quot;; gene_biotype &quot;lincRNA&quot;; transcript_name &quot;GLIS2-AS1-201&quot;; transcript_source &quot;havana&quot;; transcript_biotype &quot;lincRNA&quot;; exon_id &quot;ENSE00002667730&quot;; exon_version &quot;1&quot;; tag &quot;basic&quot;; transcript_support_level &quot;2&quot;;</span><br><span class="line">2084326:16      havana  exon    4325270 4325354 .       -       .       gene_id &quot;ENSG00000262686&quot;; gene_version &quot;1&quot;; transcript_id &quot;ENST00000576080&quot;; transcript_version &quot;1&quot;; exon_number &quot;3&quot;; gene_name &quot;GLIS2-AS1&quot;; gene_source &quot;havana&quot;; gene_biotype &quot;lincRNA&quot;; transcript_name &quot;GLIS2-AS1-201&quot;; transcript_source &quot;havana&quot;; transcript_biotype &quot;lincRNA&quot;; exon_id &quot;ENSE00002665496&quot;; exon_version &quot;1&quot;; tag &quot;basic&quot;; transcript_support_level &quot;2&quot;;</span><br><span class="line">2084327:16      havana  exon    4324667 4324775 .       -       .       gene_id &quot;ENSG00000262686&quot;; gene_version &quot;1&quot;; transcript_id &quot;ENST00000576080&quot;; transcript_version &quot;1&quot;; exon_number &quot;4&quot;; gene_name &quot;GLIS2-AS1&quot;; gene_source &quot;havana&quot;; gene_biotype &quot;lincRNA&quot;; transcript_name &quot;GLIS2-AS1-201&quot;; transcript_source &quot;havana&quot;; transcript_biotype &quot;lincRNA&quot;; exon_id &quot;ENSE00002663811&quot;; exon_version &quot;1&quot;; tag &quot;basic&quot;; transcript_support_level &quot;2&quot;;</span><br></pre></td></tr></table></figure>
<p>额，这个意思是ensembl把GLIS2-AS1这个融合基因的ENSG也列为</p>
<p>gene biotype详细解释<br><a target="_blank" rel="noopener" href="https://www.gencodegenes.org/pages/biotypes.html">https://www.gencodegenes.org/pages/biotypes.html</a></p>
<p>假基因分四类：1,经处理的假基因 (Processed)：机制是mrna反转录为DNA并继续插入染色体、通常具有poly-A尾巴，没有内含子和启动子<br>            2,未经处理的假基因 (Non-processed):基因复制后突变变为非功能性基因<br>            3,单套假基因 (Unitary pseudogenes):原本为功能性基因经过突变变为非功能性基因，保留下来。<br>            4,假的假基因 (Pseudo-pseudogenes):<br>ref:1,<a target="_blank" rel="noopener" href="https://zh.wikipedia.org/wiki/%E5%81%87%E5%9F%BA%E5%9B%A0">https://zh.wikipedia.org/wiki/%E5%81%87%E5%9F%BA%E5%9B%A0</a><br>    2,<a target="_blank" rel="noopener" href="https://highscope.ch.ntu.edu.tw/wordpress/?p=42776">https://highscope.ch.ntu.edu.tw/wordpress/?p=42776</a></p>
<p>判断一个数据框两列的关系</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">compare=<span class="keyword">lambda</span> x,y,df:<span class="built_in">len</span>([g <span class="keyword">for</span> g,df <span class="keyword">in</span> df.drop_duplicates([x,y]).groupby(x) <span class="keyword">if</span> <span class="built_in">len</span>(df[y].to_list())&gt;<span class="number">1</span>])</span><br><span class="line">compare1=<span class="keyword">lambda</span> x,y,df:[g <span class="keyword">for</span> g,df <span class="keyword">in</span> df.drop_duplicates([x,y]).groupby(x) <span class="keyword">if</span> <span class="built_in">len</span>(df[y].to_list())&gt;<span class="number">1</span>]</span><br></pre></td></tr></table></figure>

<p>统计了一下GTF文件中，大约有175个gene name对应的ENSG大于1.</p>
<p>挑选出一个基因ALG1L9P看看,<br>可以看到map了两个ENSG分别是ENSG00000248671和ENSG00000254978，</p>
<p>第一个最长的ENSG00000248671是从11号染色体的71794363到71818238，由四个transcript组成，geneType是processed_transcript（Doesn’t contain an ORF.与之对应的是protein_coding：Contains an open reading frame (ORF)）属于假基因，不编码蛋白，第二个ENSG00000254978是从染色体的71800541到71804640位置，只有一个transcript，geneType是transcribed_unprocessed_pseudogene（未经处理的假基因）。<br>由此可见，虽然这两个基因是同一个名称，但是其实是两个基因，只不过定位上第一个包含第二个。所以基因按照ENSG来。<br>再看几个个例，如：<br>基因SPATA13，他也对应两个ENSG，一个是protein_coding编码蛋白的，一个是lincRNA，长RNA非编码蛋白的基因。<br>总的来说，这些多个ENSG对应一个gene symbol，一个是对应编码蛋白的基因，一个是对应非编码蛋白的基因，而且在位置上编码蛋白的基因会包括非编码蛋白的基因。在分析蛋白表达值或者蛋白质结构复合物的时候，只分析编码蛋白的基因。</p>
<p>还有一些gene name对应特别多的ENSG id，比如：RF00443基因对应(ENSG00000200235,ENSG00000201807,ENSG00000252459,ENSG00000199474)对应四个.RF00003对应16个基因，RF02271对应36个基因，RF00019对应758个基因。<br>对于这些对应基因大于2的gene symbol，大多都是RF开头的，且gene biotype都是{‘misc_RNA’, ‘rRNA’, ‘sRNA’, ‘scaRNA’, ‘snRNA’, ‘snoRNA’}，这些都是非编码蛋白的预测的基因（Non-coding RNA predicted using sequences from Rfam and miRBase），要分析蛋白表达或者结构可以去掉。</p>
<figure class="highlight sql"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br></pre></td><td class="code"><pre><span class="line">USE GENE;</span><br><span class="line"></span><br><span class="line"><span class="keyword">CREATE</span> <span class="keyword">TABLE</span> IF <span class="keyword">NOT</span> <span class="keyword">EXISTS</span> `genes`(</span><br><span class="line">           `id` <span class="type">INT</span> UNSIGNED,</span><br><span class="line">           `GeneName` <span class="type">VARCHAR</span>(<span class="number">100</span>) <span class="keyword">NOT</span> <span class="keyword">NULL</span>,</span><br><span class="line">           `idType` ENUM(<span class="string">&#x27;ENSG&#x27;</span>,<span class="string">&#x27;synonym&#x27;</span>,<span class="string">&#x27;Entrez&#x27;</span>) <span class="keyword">NOT</span> <span class="keyword">NULL</span>,</span><br><span class="line">           `<span class="keyword">value</span>` <span class="type">VARCHAR</span>(<span class="number">100</span>)  <span class="keyword">NOT</span> <span class="keyword">NULL</span>,</span><br><span class="line">           <span class="keyword">PRIMARY</span> KEY (`id`)</span><br><span class="line">        )ENGINE<span class="operator">=</span>InnoDB <span class="keyword">DEFAULT</span> CHARSET<span class="operator">=</span>utf8;</span><br><span class="line"><span class="comment">-- INT 最大4 294967295  gtf文件注释最长位置是 248937043</span></span><br><span class="line"><span class="comment">-- ENST和ENSP和uniprot是一对一的关系，很多ENST没有对应的ENSP和uniprot，但是ENSP会对应一个ENST。</span></span><br><span class="line"><span class="comment">-- 不使用外键 查询的时候加索引</span></span><br><span class="line"></span><br><span class="line"><span class="comment">-- protein_id不为空的行，type为CDS，type为CDS的行，protein_id不为空。</span></span><br><span class="line"><span class="keyword">CREATE</span> <span class="keyword">TABLE</span> IF <span class="keyword">NOT</span> <span class="keyword">EXISTS</span> `ENSP`(</span><br><span class="line">           `id` <span class="type">VARCHAR</span>(<span class="number">200</span>) <span class="keyword">NOT</span> <span class="keyword">NULL</span>,</span><br><span class="line">           `version`  <span class="type">INT</span> UNSIGNED <span class="keyword">NOT</span> <span class="keyword">NULL</span>,</span><br><span class="line">           `ccds_id` <span class="type">VARCHAR</span>(<span class="number">100</span>) <span class="keyword">NOT</span> <span class="keyword">NULL</span>,<span class="comment">--可以是空</span></span><br><span class="line">           `chromosome` <span class="type">VARCHAR</span>(<span class="number">20</span>) <span class="keyword">NOT</span> <span class="keyword">NULL</span>,</span><br><span class="line">           `strand` ENUM(<span class="string">&#x27;+&#x27;</span>,<span class="string">&#x27;-&#x27;</span>) <span class="keyword">NOT</span> <span class="keyword">NULL</span>,</span><br><span class="line">           `<span class="keyword">start</span>` <span class="type">INT</span> UNSIGNED <span class="keyword">NOT</span> <span class="keyword">NULL</span>,</span><br><span class="line">           `<span class="keyword">end</span>` <span class="type">INT</span> UNSIGNED <span class="keyword">NOT</span> <span class="keyword">NULL</span>,</span><br><span class="line">           `ENST` <span class="type">VARCHAR</span>(<span class="number">200</span>) <span class="keyword">NOT</span> <span class="keyword">NULL</span>,</span><br><span class="line">           `ENSG` <span class="type">VARCHAR</span>(<span class="number">200</span>) <span class="keyword">NOT</span> <span class="keyword">NULL</span>,</span><br><span class="line">           <span class="keyword">PRIMARY</span> KEY (`id`)</span><br><span class="line">        )ENGINE<span class="operator">=</span>InnoDB <span class="keyword">DEFAULT</span> CHARSET<span class="operator">=</span>utf8;</span><br><span class="line"></span><br><span class="line"><span class="keyword">CREATE</span> <span class="keyword">TABLE</span> IF <span class="keyword">NOT</span> <span class="keyword">EXISTS</span> `ENST`(</span><br><span class="line">           `id` <span class="type">VARCHAR</span>(<span class="number">200</span>) <span class="keyword">NOT</span> <span class="keyword">NULL</span>,</span><br><span class="line">           `version`  <span class="type">INT</span> UNSIGNED <span class="keyword">NOT</span> <span class="keyword">NULL</span>,</span><br><span class="line">           `TranscriptName` <span class="type">VARCHAR</span>(<span class="number">100</span>) <span class="keyword">NOT</span> <span class="keyword">NULL</span>,</span><br><span class="line">           `chromosome` <span class="type">VARCHAR</span>(<span class="number">20</span>) <span class="keyword">NOT</span> <span class="keyword">NULL</span>,</span><br><span class="line">           `strand` ENUM(<span class="string">&#x27;+&#x27;</span>,<span class="string">&#x27;-&#x27;</span>) <span class="keyword">NOT</span> <span class="keyword">NULL</span>,</span><br><span class="line">           `<span class="keyword">start</span>` <span class="type">INT</span> UNSIGNED <span class="keyword">NOT</span> <span class="keyword">NULL</span>,</span><br><span class="line">           `<span class="keyword">end</span>` <span class="type">INT</span> UNSIGNED <span class="keyword">NOT</span> <span class="keyword">NULL</span>,</span><br><span class="line">           `biotype` <span class="type">VARCHAR</span>(<span class="number">50</span>) <span class="keyword">NOT</span> <span class="keyword">NULL</span>,</span><br><span class="line">           `ENSG` <span class="type">VARCHAR</span>(<span class="number">200</span>) <span class="keyword">NOT</span> <span class="keyword">NULL</span>,</span><br><span class="line">           <span class="keyword">PRIMARY</span> KEY (`id`)</span><br><span class="line">        )ENGINE<span class="operator">=</span>InnoDB <span class="keyword">DEFAULT</span> CHARSET<span class="operator">=</span>utf8;</span><br><span class="line"></span><br><span class="line"><span class="keyword">CREATE</span> <span class="keyword">TABLE</span> IF <span class="keyword">NOT</span> <span class="keyword">EXISTS</span> `ENSG`(</span><br><span class="line">           `id` <span class="type">VARCHAR</span>(<span class="number">200</span>) <span class="keyword">NOT</span> <span class="keyword">NULL</span>,</span><br><span class="line">           `version`  <span class="type">INT</span> UNSIGNED <span class="keyword">NOT</span> <span class="keyword">NULL</span>,</span><br><span class="line">           `GeneName` <span class="type">VARCHAR</span>(<span class="number">100</span>) <span class="keyword">NOT</span> <span class="keyword">NULL</span>,</span><br><span class="line">           `chromosome` <span class="type">VARCHAR</span>(<span class="number">20</span>) <span class="keyword">NOT</span> <span class="keyword">NULL</span>,</span><br><span class="line">           `strand` ENUM(<span class="string">&#x27;+&#x27;</span>,<span class="string">&#x27;-&#x27;</span>) <span class="keyword">NOT</span> <span class="keyword">NULL</span>,</span><br><span class="line">           `<span class="keyword">start</span>` <span class="type">INT</span> UNSIGNED <span class="keyword">NOT</span> <span class="keyword">NULL</span>,</span><br><span class="line">           `<span class="keyword">end</span>` <span class="type">INT</span> UNSIGNED <span class="keyword">NOT</span> <span class="keyword">NULL</span>,</span><br><span class="line">           `biotype` <span class="type">VARCHAR</span>(<span class="number">50</span>) <span class="keyword">NOT</span> <span class="keyword">NULL</span>,</span><br><span class="line">           <span class="comment">--`ENST` VARCHAR(200) NOT NULL,</span></span><br><span class="line">           <span class="keyword">PRIMARY</span> KEY (`id`),</span><br><span class="line">           <span class="comment">--FOREIGN KEY (&#x27;ENST&#x27;) REFERENCES  ENST(id)</span></span><br><span class="line">           <span class="comment">--on delete cascade</span></span><br><span class="line">           <span class="comment">--on update cascade,</span></span><br><span class="line">        )ENGINE<span class="operator">=</span>InnoDB <span class="keyword">DEFAULT</span> CHARSET<span class="operator">=</span>utf8;</span><br><span class="line"></span><br><span class="line"><span class="keyword">CREATE</span> <span class="keyword">TABLE</span> IF <span class="keyword">NOT</span> <span class="keyword">EXISTS</span> `MAP`(</span><br><span class="line">           `ENSG` <span class="type">VARCHAR</span>(<span class="number">200</span>) <span class="keyword">NOT</span> <span class="keyword">NULL</span>,</span><br><span class="line">           `ENST` <span class="type">VARCHAR</span>(<span class="number">200</span>) <span class="keyword">NOT</span> <span class="keyword">NULL</span>,</span><br><span class="line">           `ENSP` <span class="type">VARCHAR</span>(<span class="number">200</span>) <span class="keyword">NOT</span> <span class="keyword">NULL</span>,</span><br><span class="line">           `uniprot` <span class="type">VARCHAR</span>(<span class="number">200</span>) <span class="keyword">NOT</span> <span class="keyword">NULL</span>,</span><br><span class="line">           `version`  <span class="type">INT</span> UNSIGNED <span class="keyword">NOT</span> <span class="keyword">NULL</span>,</span><br><span class="line">           `TranscriptName` <span class="type">VARCHAR</span>(<span class="number">100</span>) <span class="keyword">NOT</span> <span class="keyword">NULL</span>,</span><br><span class="line">           `chromosome` <span class="type">VARCHAR</span>(<span class="number">20</span>) <span class="keyword">NOT</span> <span class="keyword">NULL</span>,</span><br><span class="line">           `strand` ENUM(<span class="string">&#x27;+&#x27;</span>,<span class="string">&#x27;-&#x27;</span>) <span class="keyword">NOT</span> <span class="keyword">NULL</span>,</span><br><span class="line">           `<span class="keyword">start</span>` <span class="type">INT</span> UNSIGNED <span class="keyword">NOT</span> <span class="keyword">NULL</span>,</span><br><span class="line">           `<span class="keyword">end</span>` <span class="type">INT</span> UNSIGNED <span class="keyword">NOT</span> <span class="keyword">NULL</span>,</span><br><span class="line">           <span class="keyword">PRIMARY</span> KEY (`id`)</span><br><span class="line">        )ENGINE<span class="operator">=</span>InnoDB <span class="keyword">DEFAULT</span> CHARSET<span class="operator">=</span>utf8;</span><br><span class="line"></span><br><span class="line"><span class="comment">-- ON UPDATE CASCADE</span></span><br><span class="line"><span class="comment">-- ON DELETE RESTRICT</span></span><br><span class="line"></span><br><span class="line"><span class="comment">-- 外键约束参数</span></span><br><span class="line"><span class="comment">-- CASCADE——删除包含与已删除键值，有参照关系的所有记录</span></span><br><span class="line"><span class="comment">-- SET NULL——修改包含与已删除键值，有参照关系的所有记录，使用NULL值替换，不能用于已标记为NOT NULL的字段</span></span><br><span class="line"><span class="comment">-- NO ACTION——不进行任何操作</span></span><br><span class="line"><span class="comment">-- RESTRICT——拒绝主表删除或修改外键关联词，在不定义ON DELETE和ON UPDATE子句时，这是默认设置，也是最安全的的设置</span></span><br><span class="line"></span><br></pre></td></tr></table></figure>




<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br><span class="line">90</span><br><span class="line">91</span><br><span class="line">92</span><br><span class="line">93</span><br><span class="line">94</span><br><span class="line">95</span><br><span class="line">96</span><br><span class="line">97</span><br><span class="line">98</span><br><span class="line">99</span><br><span class="line">100</span><br><span class="line">101</span><br><span class="line">102</span><br><span class="line">103</span><br><span class="line">104</span><br><span class="line">105</span><br><span class="line">106</span><br><span class="line">107</span><br><span class="line">108</span><br><span class="line">109</span><br><span class="line">110</span><br><span class="line">111</span><br><span class="line">112</span><br><span class="line">113</span><br><span class="line">114</span><br><span class="line">115</span><br><span class="line">116</span><br><span class="line">117</span><br><span class="line">118</span><br><span class="line">119</span><br><span class="line">120</span><br><span class="line">121</span><br><span class="line">122</span><br><span class="line">123</span><br><span class="line">124</span><br><span class="line">125</span><br><span class="line">126</span><br><span class="line">127</span><br><span class="line">128</span><br><span class="line">129</span><br><span class="line">130</span><br><span class="line">131</span><br><span class="line">132</span><br><span class="line">133</span><br><span class="line">134</span><br><span class="line">135</span><br><span class="line">136</span><br><span class="line">137</span><br><span class="line">138</span><br><span class="line">139</span><br><span class="line">140</span><br><span class="line">141</span><br><span class="line">142</span><br><span class="line">143</span><br><span class="line">144</span><br><span class="line">145</span><br><span class="line">146</span><br><span class="line">147</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">import</span> logging</span><br><span class="line">logging.basicConfig(<span class="built_in">format</span>=<span class="string">&#x27;%(asctime)s - %(pathname)s[line:%(lineno)d] - %(levelname)s: %(message)s&#x27;</span>,</span><br><span class="line">                    level=logging.INFO)</span><br><span class="line"><span class="keyword">import</span> sqlite3</span><br><span class="line">dbfile=<span class="string">&#x27;/data/wenyuhao/semiLearn/TMP/EN.db&#x27;</span></span><br><span class="line">logging.info(<span class="string">f&#x27;init DB file : <span class="subst">&#123;dbfile&#125;</span>&#x27;</span>)</span><br><span class="line">conn = sqlite3.connect(dbfile)</span><br><span class="line">c = conn.cursor()</span><br><span class="line"></span><br><span class="line">logging.info(<span class="string">f&#x27;setting...&#x27;</span>)</span><br><span class="line">c.execute(<span class="string">&#x27;&#x27;&#x27;--sqlite 建立数据库</span></span><br><span class="line"><span class="string">-- sqlite没有枚举类型，将枚举类型转化为5个长度VARCHAR类型</span></span><br><span class="line"><span class="string">PRAGMA encoding = &quot;UTF-8&quot;;</span></span><br><span class="line"><span class="string">&#x27;&#x27;&#x27;</span>)</span><br><span class="line"></span><br><span class="line">logging.info(<span class="string">f&#x27;creat table CDS&#x27;</span>)</span><br><span class="line">c.execute(</span><br><span class="line"><span class="string">&#x27;&#x27;&#x27;</span></span><br><span class="line"><span class="string">CREATE TABLE IF NOT EXISTS `CDS`(--几个CDS组成一个ENSP</span></span><br><span class="line"><span class="string">           `id` VARCHAR(400) NOT NULL,--uuid作为主键</span></span><br><span class="line"><span class="string">           `ccds_id` VARCHAR(100),--可以是空</span></span><br><span class="line"><span class="string">           `chromosome` VARCHAR(20) NOT NULL,</span></span><br><span class="line"><span class="string">           `strand` VARCHAR(5) NOT NULL,</span></span><br><span class="line"><span class="string">           `start` INT UNSIGNED NOT NULL,</span></span><br><span class="line"><span class="string">           `end` INT UNSIGNED NOT NULL,</span></span><br><span class="line"><span class="string">           `ENSP` VARCHAR(200) NOT NULL,</span></span><br><span class="line"><span class="string">           `ENST` VARCHAR(200) NOT NULL,--一个ENSP是对应一个ENSG，也可能没有对应任何ENST</span></span><br><span class="line"><span class="string">           `ENSG` VARCHAR(200) NOT NULL,--一个ENSG对应多个ENSP  一个ENSP对应一个ENSG</span></span><br><span class="line"><span class="string">           PRIMARY KEY (`id`)</span></span><br><span class="line"><span class="string">        );</span></span><br><span class="line"><span class="string">&#x27;&#x27;&#x27;</span>)</span><br><span class="line">logging.info(<span class="string">f&#x27;creat table ENSP&#x27;</span>)</span><br><span class="line">c.execute(</span><br><span class="line"><span class="string">&#x27;&#x27;&#x27;</span></span><br><span class="line"><span class="string">CREATE TABLE IF NOT EXISTS `ENSP`(</span></span><br><span class="line"><span class="string">           `id` VARCHAR(200) NOT NULL,</span></span><br><span class="line"><span class="string">           `version`  INT UNSIGNED NOT NULL,</span></span><br><span class="line"><span class="string">           --`chromosome` VARCHAR(20) NOT NULL,</span></span><br><span class="line"><span class="string">           --`strand` VARCHAR(5) NOT NULL,</span></span><br><span class="line"><span class="string">           --`start` INT UNSIGNED NOT NULL,--没有start，第一个CDS的start</span></span><br><span class="line"><span class="string">           --`end` INT UNSIGNED NOT NULL,--没有end，最后一个CDS的end</span></span><br><span class="line"><span class="string">           `ENST` VARCHAR(200) NOT NULL,--一个ENSP是对应一个ENSG，也可能没有对应任何ENST</span></span><br><span class="line"><span class="string">           `ENSG` VARCHAR(200) NOT NULL,--一个ENSG对应多个ENSP  一个ENSP对应一个ENSG</span></span><br><span class="line"><span class="string">           PRIMARY KEY (`id`)</span></span><br><span class="line"><span class="string">        );</span></span><br><span class="line"><span class="string">&#x27;&#x27;&#x27;</span>)</span><br><span class="line">logging.info(<span class="string">f&#x27;creat table ENST&#x27;</span>)</span><br><span class="line">c.execute(<span class="string">&#x27;&#x27;&#x27;</span></span><br><span class="line"><span class="string">CREATE TABLE IF NOT EXISTS `ENST`(</span></span><br><span class="line"><span class="string">           `id` VARCHAR(200) NOT NULL,</span></span><br><span class="line"><span class="string">           `version`  INT UNSIGNED NOT NULL,</span></span><br><span class="line"><span class="string">           `TranscriptName` VARCHAR(100) NOT NULL,</span></span><br><span class="line"><span class="string">           `chromosome` VARCHAR(20) NOT NULL,</span></span><br><span class="line"><span class="string">           `strand` VARCHAR(5) NOT NULL,</span></span><br><span class="line"><span class="string">           `start` INT UNSIGNED NOT NULL,</span></span><br><span class="line"><span class="string">           `end` INT UNSIGNED NOT NULL,</span></span><br><span class="line"><span class="string">           `biotype` VARCHAR(50) NOT NULL,</span></span><br><span class="line"><span class="string">           `ENSG` VARCHAR(200) NOT NULL,</span></span><br><span class="line"><span class="string">           PRIMARY KEY (`id`)</span></span><br><span class="line"><span class="string">        );</span></span><br><span class="line"><span class="string">&#x27;&#x27;&#x27;</span>)</span><br><span class="line">logging.info(<span class="string">f&#x27;creat table ENSG&#x27;</span>)</span><br><span class="line">c.execute(<span class="string">&#x27;&#x27;&#x27;</span></span><br><span class="line"><span class="string">CREATE TABLE IF NOT EXISTS `ENSG`(</span></span><br><span class="line"><span class="string">           `id` VARCHAR(200) NOT NULL,</span></span><br><span class="line"><span class="string">           `version`  INT UNSIGNED NOT NULL,</span></span><br><span class="line"><span class="string">           `GeneName` VARCHAR(100) NOT NULL,</span></span><br><span class="line"><span class="string">           `chromosome` VARCHAR(20) NOT NULL,</span></span><br><span class="line"><span class="string">           `strand` VARCHAR(5) NOT NULL,</span></span><br><span class="line"><span class="string">           `start` INT UNSIGNED NOT NULL,</span></span><br><span class="line"><span class="string">           `end` INT UNSIGNED NOT NULL,</span></span><br><span class="line"><span class="string">           `biotype` VARCHAR(50) NOT NULL,</span></span><br><span class="line"><span class="string">            PRIMARY KEY (`id`)</span></span><br><span class="line"><span class="string">        );</span></span><br><span class="line"><span class="string">&#x27;&#x27;&#x27;</span>)</span><br><span class="line">logging.info(<span class="string">f&#x27;creat table ENSE&#x27;</span>)</span><br><span class="line">c.execute(<span class="string">&#x27;&#x27;&#x27;--ENSE和ENST是多对多 ENSE和ENSG是一对多,所以ENST不算在属性内部</span></span><br><span class="line"><span class="string">CREATE TABLE IF NOT EXISTS `ENSE`(</span></span><br><span class="line"><span class="string">           `id` VARCHAR(200) NOT NULL,</span></span><br><span class="line"><span class="string">           `version`  INT UNSIGNED NOT NULL,</span></span><br><span class="line"><span class="string">           `chromosome` VARCHAR(20) NOT NULL,</span></span><br><span class="line"><span class="string">           `strand` VARCHAR(5) NOT NULL,</span></span><br><span class="line"><span class="string">           `start` INT UNSIGNED NOT NULL,</span></span><br><span class="line"><span class="string">           `end` INT UNSIGNED NOT NULL,</span></span><br><span class="line"><span class="string">           `ENSG` VARCHAR(200) NOT NULL,</span></span><br><span class="line"><span class="string">            PRIMARY KEY (`id`)</span></span><br><span class="line"><span class="string">        );</span></span><br><span class="line"><span class="string">&#x27;&#x27;&#x27;</span>)</span><br><span class="line">conn.commit()</span><br><span class="line"></span><br><span class="line">logging.info(<span class="string">f&#x27;插入ENSG的数据&#x27;</span>)</span><br><span class="line"><span class="comment">#插入ENSG的数据</span></span><br><span class="line">sql=<span class="string">&#x27;INSERT INTO ENSG (id,version,GeneName,chromosome,strand,start,end,biotype) VALUES &#x27;</span></span><br><span class="line">sql+=<span class="string">&#x27;,&#x27;</span>.join([<span class="string">f&quot;(&#x27;<span class="subst">&#123;x[<span class="string">&#x27;gene_id&#x27;</span>]&#125;</span>&#x27;,<span class="subst">&#123;x[<span class="string">&#x27;gene_version&#x27;</span>]&#125;</span>,&#x27;<span class="subst">&#123;x[<span class="string">&#x27;gene_name&#x27;</span>]&#125;</span>&#x27;,&#x27;<span class="subst">&#123;x[<span class="string">&#x27;chr&#x27;</span>]&#125;</span>&#x27;,&#x27;<span class="subst">&#123;x[<span class="string">&#x27;strand&#x27;</span>]&#125;</span>&#x27;,<span class="subst">&#123;x[<span class="string">&#x27;start&#x27;</span>]&#125;</span>,<span class="subst">&#123;x[<span class="string">&#x27;end&#x27;</span>]&#125;</span>,&#x27;<span class="subst">&#123;x[<span class="string">&#x27;gene_biotype&#x27;</span>]&#125;</span>&#x27;)&quot;</span> <span class="keyword">for</span> _,x <span class="keyword">in</span> sha1.query(<span class="string">&#x27;type==&quot;gene&quot;&#x27;</span>).iterrows()])+<span class="string">&quot; ;&quot;</span></span><br><span class="line">c.execute(sql)</span><br><span class="line">conn.commit()</span><br><span class="line"></span><br><span class="line">logging.info(<span class="string">f&#x27;插入ENST的数据&#x27;</span>)</span><br><span class="line"><span class="comment">#插入ENST的数据</span></span><br><span class="line">sql=<span class="string">&#x27;INSERT INTO ENST (id,version,TranscriptName,chromosome,strand,start,end,biotype,ENSG) VALUES &#x27;</span></span><br><span class="line">sql+=<span class="string">&#x27;,&#x27;</span>.join([<span class="string">f&quot;(&#x27;<span class="subst">&#123;x[<span class="string">&#x27;transcript_id&#x27;</span>]&#125;</span>&#x27;,<span class="subst">&#123;x[<span class="string">&#x27;transcript_version&#x27;</span>]&#125;</span>,&#x27;<span class="subst">&#123;x[<span class="string">&#x27;transcript_name&#x27;</span>]&#125;</span>&#x27;,&#x27;<span class="subst">&#123;x[<span class="string">&#x27;chr&#x27;</span>]&#125;</span>&#x27;,&#x27;<span class="subst">&#123;x[<span class="string">&#x27;strand&#x27;</span>]&#125;</span>&#x27;,<span class="subst">&#123;x[<span class="string">&#x27;start&#x27;</span>]&#125;</span>,<span class="subst">&#123;x[<span class="string">&#x27;end&#x27;</span>]&#125;</span>,&#x27;<span class="subst">&#123;x[<span class="string">&#x27;transcript_biotype&#x27;</span>]&#125;</span>&#x27;,&#x27;<span class="subst">&#123;x[<span class="string">&#x27;gene_id&#x27;</span>]&#125;</span>&#x27;)&quot;</span> <span class="keyword">for</span> _,x <span class="keyword">in</span> sha1.query(<span class="string">&#x27;type==&quot;transcript&quot;&#x27;</span>).iterrows()])+<span class="string">&quot; ;&quot;</span></span><br><span class="line">c.execute(sql)</span><br><span class="line">conn.commit()</span><br><span class="line"></span><br><span class="line">logging.info(<span class="string">f&#x27;插入ENSE的数据&#x27;</span>)</span><br><span class="line"><span class="comment">#插入ENSE的数据</span></span><br><span class="line">sql=<span class="string">&#x27;INSERT INTO ENSE (id,version,chromosome,strand,start,end,ENSG) VALUES &#x27;</span></span><br><span class="line">sql+=<span class="string">&#x27;,&#x27;</span>.join([<span class="string">f&quot;(&#x27;<span class="subst">&#123;x[<span class="string">&#x27;exon_id&#x27;</span>]&#125;</span>&#x27;,<span class="subst">&#123;x[<span class="string">&#x27;exon_version&#x27;</span>]&#125;</span>,&#x27;<span class="subst">&#123;x[<span class="string">&#x27;chr&#x27;</span>]&#125;</span>&#x27;,&#x27;<span class="subst">&#123;x[<span class="string">&#x27;strand&#x27;</span>]&#125;</span>&#x27;,<span class="subst">&#123;x[<span class="string">&#x27;start&#x27;</span>]&#125;</span>,<span class="subst">&#123;x[<span class="string">&#x27;end&#x27;</span>]&#125;</span>,&#x27;<span class="subst">&#123;x[<span class="string">&#x27;gene_id&#x27;</span>]&#125;</span>&#x27;)&quot;</span> <span class="keyword">for</span> _,x <span class="keyword">in</span> sha1.query(<span class="string">&#x27;type==&quot;exon&quot;&#x27;</span>).iterrows()])+<span class="string">&quot; ;&quot;</span></span><br><span class="line">c.execute(sql)</span><br><span class="line">conn.commit()</span><br><span class="line"></span><br><span class="line">logging.info(<span class="string">f&#x27;#插入CDS的数据&#x27;</span>)</span><br><span class="line"><span class="comment">#插入CDS的数据</span></span><br><span class="line"><span class="keyword">import</span> uuid</span><br><span class="line">sql=<span class="string">&#x27;INSERT INTO CDS (id,ccds_id,chromosome,strand,start,end,ENSP,ENST,ENSG) VALUES &#x27;</span></span><br><span class="line">tmp=sha1.query(<span class="string">&#x27;type==&quot;CDS&quot;&#x27;</span>)</span><br><span class="line">tmp[<span class="string">&#x27;ccds_id&#x27;</span>]=tmp[<span class="string">&#x27;ccds_id&#x27;</span>].apply(<span class="keyword">lambda</span> x:<span class="string">&#x27;NULL&#x27;</span> <span class="keyword">if</span> pd.isna(x) <span class="keyword">else</span>  <span class="string">f&quot;&#x27;<span class="subst">&#123;x&#125;</span>&#x27;&quot;</span>)</span><br><span class="line">sql+=<span class="string">&#x27;,&#x27;</span>.join([<span class="string">f&quot;(&#x27;<span class="subst">&#123;uuid.uuid4().__str__()&#125;</span>&#x27;,<span class="subst">&#123;x[<span class="string">&#x27;ccds_id&#x27;</span>]&#125;</span>,&#x27;<span class="subst">&#123;x[<span class="string">&#x27;chr&#x27;</span>]&#125;</span>&#x27;,&#x27;<span class="subst">&#123;x[<span class="string">&#x27;strand&#x27;</span>]&#125;</span>&#x27;,<span class="subst">&#123;x[<span class="string">&#x27;start&#x27;</span>]&#125;</span>,<span class="subst">&#123;x[<span class="string">&#x27;end&#x27;</span>]&#125;</span>,&#x27;<span class="subst">&#123;x[<span class="string">&#x27;protein_id&#x27;</span>]&#125;</span>&#x27;,&#x27;<span class="subst">&#123;x[<span class="string">&#x27;transcript_id&#x27;</span>]&#125;</span>&#x27;,&#x27;<span class="subst">&#123;x[<span class="string">&#x27;gene_id&#x27;</span>]&#125;</span>&#x27;)&quot;</span> <span class="keyword">for</span> _,x <span class="keyword">in</span> tmp.iterrows()])+<span class="string">&quot; ;&quot;</span></span><br><span class="line">c.execute(sql)</span><br><span class="line">conn.commit()</span><br><span class="line"></span><br><span class="line">logging.info(<span class="string">f&#x27;插入ENSP的数据&#x27;</span>)</span><br><span class="line"><span class="comment">#插入ENSP的数据</span></span><br><span class="line">sql=<span class="string">&#x27;INSERT INTO ENSP (id,version,ENST,ENSG) VALUES &#x27;</span></span><br><span class="line">sql+=<span class="string">&#x27;,&#x27;</span>.join([<span class="string">f&quot;(&#x27;<span class="subst">&#123;x[<span class="string">&#x27;protein_id&#x27;</span>]&#125;</span>&#x27;,<span class="subst">&#123;x[<span class="string">&#x27;protein_version&#x27;</span>]&#125;</span>,&#x27;<span class="subst">&#123;x[<span class="string">&#x27;transcript_id&#x27;</span>]&#125;</span>&#x27;,&#x27;<span class="subst">&#123;x[<span class="string">&#x27;gene_id&#x27;</span>]&#125;</span>&#x27;)&quot;</span> <span class="keyword">for</span> _,x <span class="keyword">in</span> sha1.query(<span class="string">&#x27;type==&quot;CDS&quot;&#x27;</span>).drop_duplicates(<span class="string">&#x27;protein_id&#x27;</span>).iterrows()])+<span class="string">&quot; ;&quot;</span></span><br><span class="line">c.execute(sql)</span><br><span class="line">conn.commit()</span><br><span class="line"></span><br><span class="line">conn.close()</span><br><span class="line"></span><br><span class="line"><span class="comment">#插入语句</span></span><br><span class="line">sql=<span class="string">&#x27;USE GENE;\nINSERT INTO ENSG (id,version,GeneName,chromosome,strand,start,end,biotype) VALUES &#x27;</span></span><br><span class="line">sql+=<span class="string">&#x27;,&#x27;</span>.join([<span class="string">f&quot;(&#x27;<span class="subst">&#123;x[<span class="string">&#x27;gene_id&#x27;</span>]&#125;</span>&#x27;,<span class="subst">&#123;x[<span class="string">&#x27;gene_version&#x27;</span>]&#125;</span>,&#x27;<span class="subst">&#123;x[<span class="string">&#x27;gene_name&#x27;</span>]&#125;</span>&#x27;,&#x27;<span class="subst">&#123;x[<span class="string">&#x27;chr&#x27;</span>]&#125;</span>&#x27;,&#x27;<span class="subst">&#123;x[<span class="string">&#x27;strand&#x27;</span>]&#125;</span>&#x27;,<span class="subst">&#123;x[<span class="string">&#x27;start&#x27;</span>]&#125;</span>,<span class="subst">&#123;x[<span class="string">&#x27;end&#x27;</span>]&#125;</span>,&#x27;<span class="subst">&#123;x[<span class="string">&#x27;gene_biotype&#x27;</span>]&#125;</span>&#x27;)&quot;</span> <span class="keyword">for</span> _,x <span class="keyword">in</span> sha1.query(<span class="string">&#x27;type==&quot;gene&quot;&#x27;</span>).iterrows()])+<span class="string">&quot; ;&quot;</span></span><br><span class="line">sql+=<span class="string">&#x27;INSERT INTO ENST (id,version,TranscriptName,chromosome,strand,start,end,biotype,ENSG) VALUES &#x27;</span></span><br><span class="line">sql+=<span class="string">&#x27;,&#x27;</span>.join([<span class="string">f&quot;(&#x27;<span class="subst">&#123;x[<span class="string">&#x27;transcript_id&#x27;</span>]&#125;</span>&#x27;,<span class="subst">&#123;x[<span class="string">&#x27;transcript_version&#x27;</span>]&#125;</span>,&#x27;<span class="subst">&#123;x[<span class="string">&#x27;transcript_name&#x27;</span>]&#125;</span>&#x27;,&#x27;<span class="subst">&#123;x[<span class="string">&#x27;chr&#x27;</span>]&#125;</span>&#x27;,&#x27;<span class="subst">&#123;x[<span class="string">&#x27;strand&#x27;</span>]&#125;</span>&#x27;,<span class="subst">&#123;x[<span class="string">&#x27;start&#x27;</span>]&#125;</span>,<span class="subst">&#123;x[<span class="string">&#x27;end&#x27;</span>]&#125;</span>,&#x27;<span class="subst">&#123;x[<span class="string">&#x27;transcript_biotype&#x27;</span>]&#125;</span>&#x27;,&#x27;<span class="subst">&#123;x[<span class="string">&#x27;gene_id&#x27;</span>]&#125;</span>&#x27;)&quot;</span> <span class="keyword">for</span> _,x <span class="keyword">in</span> sha1.query(<span class="string">&#x27;type==&quot;transcript&quot;&#x27;</span>).iterrows()])+<span class="string">&quot; ;\n&quot;</span></span><br><span class="line">sql+=<span class="string">&#x27;INSERT INTO ENSE (id,version,chromosome,strand,start,end,ENSG) VALUES &#x27;</span></span><br><span class="line">sql+=<span class="string">&#x27;,&#x27;</span>.join([<span class="string">f&quot;(&#x27;<span class="subst">&#123;x[<span class="string">&#x27;exon_id&#x27;</span>]&#125;</span>&#x27;,<span class="subst">&#123;x[<span class="string">&#x27;exon_version&#x27;</span>]&#125;</span>,&#x27;<span class="subst">&#123;x[<span class="string">&#x27;chr&#x27;</span>]&#125;</span>&#x27;,&#x27;<span class="subst">&#123;x[<span class="string">&#x27;strand&#x27;</span>]&#125;</span>&#x27;,<span class="subst">&#123;x[<span class="string">&#x27;start&#x27;</span>]&#125;</span>,<span class="subst">&#123;x[<span class="string">&#x27;end&#x27;</span>]&#125;</span>,\</span></span><br><span class="line"><span class="string">&#x27;<span class="subst">&#123;x[<span class="string">&#x27;gene_id&#x27;</span>]&#125;</span>&#x27;)&quot;</span> <span class="keyword">for</span> _,x <span class="keyword">in</span> sha1.query(<span class="string">&#x27;type==&quot;exon&quot;&#x27;</span>).drop_duplicates(<span class="string">&#x27;exon_id&#x27;</span>).iterrows()])+<span class="string">&quot; ;\n&quot;</span></span><br><span class="line">sql+=<span class="string">&#x27;INSERT INTO CDS (id,ccds_id,chromosome,strand,start,end,ENSP,ENST,ENSG) VALUES &#x27;</span></span><br><span class="line">tmp=sha1.query(<span class="string">&#x27;type==&quot;CDS&quot;&#x27;</span>)</span><br><span class="line">tmp[<span class="string">&#x27;ccds_id&#x27;</span>]=tmp[<span class="string">&#x27;ccds_id&#x27;</span>].apply(<span class="keyword">lambda</span> x:<span class="string">&#x27;NULL&#x27;</span> <span class="keyword">if</span> pd.isna(x) <span class="keyword">else</span>  <span class="string">f&quot;&#x27;<span class="subst">&#123;x&#125;</span>&#x27;&quot;</span>)</span><br><span class="line">sql+=<span class="string">&#x27;,&#x27;</span>.join([<span class="string">f&quot;(&#x27;<span class="subst">&#123;uuid.uuid4().__str__()&#125;</span>&#x27;,<span class="subst">&#123;x[<span class="string">&#x27;ccds_id&#x27;</span>]&#125;</span>,&#x27;<span class="subst">&#123;x[<span class="string">&#x27;chr&#x27;</span>]&#125;</span>&#x27;,&#x27;<span class="subst">&#123;x[<span class="string">&#x27;strand&#x27;</span>]&#125;</span>&#x27;,<span class="subst">&#123;x[<span class="string">&#x27;start&#x27;</span>]&#125;</span>,<span class="subst">&#123;x[<span class="string">&#x27;end&#x27;</span>]&#125;</span>,&#x27;<span class="subst">&#123;x[<span class="string">&#x27;protein_id&#x27;</span>]&#125;</span>&#x27;,&#x27;<span class="subst">&#123;x[<span class="string">&#x27;transcript_id&#x27;</span>]&#125;</span>&#x27;,&#x27;<span class="subst">&#123;x[<span class="string">&#x27;gene_id&#x27;</span>]&#125;</span>&#x27;)&quot;</span> <span class="keyword">for</span> _,x <span class="keyword">in</span> tmp.iterrows()])+<span class="string">&quot; ;\n&quot;</span></span><br><span class="line">sql+=<span class="string">&#x27;INSERT INTO ENSP (id,version,ENST,ENSG) VALUES &#x27;</span></span><br><span class="line">sql+=<span class="string">&#x27;,&#x27;</span>.join([<span class="string">f&quot;(&#x27;<span class="subst">&#123;x[<span class="string">&#x27;protein_id&#x27;</span>]&#125;</span>&#x27;,<span class="subst">&#123;x[<span class="string">&#x27;protein_version&#x27;</span>]&#125;</span>,&#x27;<span class="subst">&#123;x[<span class="string">&#x27;transcript_id&#x27;</span>]&#125;</span>&#x27;,&#x27;<span class="subst">&#123;x[<span class="string">&#x27;gene_id&#x27;</span>]&#125;</span>&#x27;)&quot;</span> <span class="keyword">for</span> _,x <span class="keyword">in</span> sha1.query(<span class="string">&#x27;type==&quot;CDS&quot;&#x27;</span>).drop_duplicates(<span class="string">&#x27;protein_id&#x27;</span>).iterrows()])+<span class="string">&quot; ;\n&quot;</span></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br></pre></td></tr></table></figure>
      </section>

      
      

      
      
    </div>
    <div class="widget-info">
      <section class="widget-author widget-item layout-margin content-padding--primary soft-size--large soft-style--box">
  <div class="widget-body">
    
      <img src="https://avatars.githubusercontent.com/u/40107519?v=4" class="soft-size--round soft-style--box" alt="wenyuhao">
    
    
      <h2>wenyuhao</h2>
    
    
      <p>生物信息学</p>
    

    <div class="count-box">
      <div class="count-box--item">
        <svg class="icon icon-article" viewBox="0 0 1024 1024" version="1.1" xmlns="http://www.w3.org/2000/svg">
  <path d="M240.51564747 647.74217627h196.07203239c16.59071043 0 30.16492806-13.57421762 30.16492805-30.16492806V165.10332731c0-33.18142087-30.16492806-60.32985613-60.32985612-60.32985611H245.04038668C225.43318342 104.7734712 210.35071939 119.85593522 210.35071939 139.46313845V617.57724821c0 16.59071043 13.57421762 30.16492806 30.16492808 30.16492806z m663.62841731-452.47392089v482.63884894c0 33.18142087-27.14843525 60.32985613-60.32985612 60.32985613H180.18579134c-33.18142087 0-60.32985613-27.14843525-60.32985612-60.32985613V195.26825538c-49.77213131 0-90.49478418 40.72265287-90.49478417 90.49478417v452.4739209c0 49.77213131 40.72265287 90.49478418 90.49478417 90.49478417h286.56681657c16.59071043 0 30.16492806 13.57421762 30.16492807 30.16492807s13.57421762 30.16492806 30.16492805 30.16492806h90.49478418c16.59071043 0 30.16492806-13.57421762 30.16492805-30.16492806s13.57421762-30.16492806 30.16492807-30.16492807h286.56681657c49.77213131 0 90.49478418-40.72265287 90.49478417-90.49478417V285.76303955c0-49.77213131-40.72265287-90.49478418-90.49478417-90.49478417zM587.41232014 647.74217627h191.54729318c19.60720323 0 34.68966726-15.08246403 34.68966729-34.68966727V134.93839925c0-16.59071043-13.57421762-30.16492806-30.16492808-30.16492805H617.57724821c-30.16492806 0-60.32985613 27.14843525-60.32985612 60.32985611v452.4739209c0 16.59071043 13.57421762 30.16492806 30.16492805 30.16492806z" fill="currentColor"></path>
</svg>
        <span>11</span>
      </div>
      <div class="count-box--item">
        <svg class="icon icon-categories" viewBox="0 0 1024 1024" version="1.1" xmlns="http://www.w3.org/2000/svg">
  <path d="M900.3614811 257.09082106h-339.81629553l-67.96326003-101.9448889c-19.41807444-29.12711113-48.54518557-43.69066667-82.52681443-43.69066667H123.6385189c-53.39970333 0-97.09036999 43.69066667-97.09037113 97.09036999v582.54222222c0 53.39970333 43.69066667 97.09036999 97.09037113 97.09037002h776.7229622c53.39970333 0 97.09036999-43.69066667 97.09037113-97.09037002V354.18119104c0-53.39970333-43.69066667-97.09036999-97.09037113-97.09036998z m-97.09036999 242.72592554H220.72888889c-24.27259221 0-48.54518557-24.27259221-48.54518556-48.54518556s24.27259221-48.54518557 48.54518556-48.54518444h582.54222222c24.27259221 0 48.54518557 24.27259221 48.54518556 48.54518444s-24.27259221 48.54518557-48.54518556 48.54518556z" fill="currentColor"></path>
</svg>
        5
      </div>
      <div class="count-box--item">
        <svg class="icon icon-tags" viewBox="0 0 1098 1024" version="1.1" xmlns="http://www.w3.org/2000/svg">
  <path d="M283.42180005 272q0-28.38857157-20.09142843-48.48000001t-48.47999998-20.09142842-48.48000002 20.09142842-20.09142846 48.48000001 20.09142846 48.48 48.48000002 20.09142843 48.47999998-20.09142843 20.09142843-48.48zM855.0332285 580.57142843q0 28.38857157-19.81714313 48.2057147l-263.03999997 263.58857157q-20.9142853 19.81714313-48.75428534 19.81714312-28.38857157 0-48.20571468-19.81714312l-383.04-383.58857157q-20.36571468-19.81714313-34.55999999-54.10285688t-14.19428534-62.6742853l0-222.85714313q0-27.84000002 20.36571469-48.20571469t48.2057147-20.36571466l222.85714313 0q28.38857157 0 62.6742853 14.19428529t54.65142842 34.55999999l383.04000001 382.49142843q19.81714313 20.9142853 19.81714314 48.75428532zM1060.74751475 580.57142843q0 28.38857157-19.81714313 48.2057147l-263.04 263.58857157q-20.9142853 19.81714313-48.75428531 19.81714312-19.26857155 0-31.61142843-7.47428531t-28.38857159-24.13714314l251.79428534-251.7942853q19.81714313-19.81714313 19.81714308-48.20571469 0-27.84000002-19.81714308-48.75428531l-383.04000001-382.49142845q-20.36571468-20.36571468-54.65142842-34.55999999t-62.67428532-14.19428534l120 0q28.38857157 0 62.67428532 14.19428534t54.65142842 34.55999999l383.03999998 382.49142845q19.81714313 20.9142853 19.81714314 48.75428531z" fill="currentColor"></path>
</svg>
        2
      </div>
    </div>
  </div>
</section>

      
<section class="widet-notice widget-item layout-margin content-padding--primary soft-size--large soft-style--box">
  <div class="widget-title">
    <svg class="icon icon-notice" viewBox="0 0 1024 1024" version="1.1" xmlns="http://www.w3.org/2000/svg">
  <path d="M512 945.02305225v28.15620663a24.27259221 24.27259221 0 0 1-24.27259221 24.27259335H394.0352a48.54518557 48.54518557 0 0 1-41.74885888-23.78714112l-110.68302222-184.47170332a132.04290333 132.04290333 0 0 1-17.47626667-48.54518557h118.4502511a200.97706667 200.97706667 0 0 1 76.21594113 14.56355556l20.38897777 133.49925888a48.54518557 48.54518557 0 0 0 36.40888888 27.67075555l16.01991111 2.91271112a24.27259221 24.27259221 0 0 1 20.38897778 25.72894889zM997.45185223 463.45481443a194.18074112 194.18074112 0 0 1-38.8361489 116.50844445 24.75804445 24.75804445 0 0 1-36.4088889 0l-34.95253333-34.95253333a24.27259221 24.27259221 0 0 1-2.91271111-30.58346667 97.09036999 97.09036999 0 0 0 0-106.79940665 24.27259221 24.27259221 0 0 1 2.91271111-30.58346666l34.95253333-34.95253334a24.75804445 24.75804445 0 0 1 18.93262223-7.28177777 26.2144 26.2144 0 0 1 17.47626667 9.70903665A194.18074112 194.18074112 0 0 1 997.45185223 463.45481443z m-194.18074112-388.36148111v776.72296335a48.54518557 48.54518557 0 0 1-48.54518556 48.54518443h-28.64165888a48.54518557 48.54518557 0 0 1-33.98163001-14.07810332l-145.63555556-143.20829668A291.27111111 291.27111111 0 0 0 342.57730333 657.63555556H172.18370333a145.63555556 145.63555556 0 0 1-145.63555556-145.63555556v-97.09036999a145.63555556 145.63555556 0 0 1 145.63555556-145.63555556h170.3936a291.27111111 291.27111111 0 0 0 206.31703779-85.43952668l145.63555555-143.20829554a48.54518557 48.54518557 0 0 1 33.98162888-14.07810446H754.72592555a48.54518557 48.54518557 0 0 1 48.54518556 48.54518555z" fill="currentColor"></path>
</svg>
    <span>NOTICE</span>
  </div>
  <div class="widget-body">
    <p>qq 3484905018</p>
  </div>
</section>


      <section class="widget-categorys widget-item layout-margin content-padding--primary soft-size--large soft-style--box">
  <div class="widget-title">
    <svg class="icon icon-categories" viewBox="0 0 1024 1024" version="1.1" xmlns="http://www.w3.org/2000/svg">
  <path d="M900.3614811 257.09082106h-339.81629553l-67.96326003-101.9448889c-19.41807444-29.12711113-48.54518557-43.69066667-82.52681443-43.69066667H123.6385189c-53.39970333 0-97.09036999 43.69066667-97.09037113 97.09036999v582.54222222c0 53.39970333 43.69066667 97.09036999 97.09037113 97.09037002h776.7229622c53.39970333 0 97.09036999-43.69066667 97.09037113-97.09037002V354.18119104c0-53.39970333-43.69066667-97.09036999-97.09037113-97.09036998z m-97.09036999 242.72592554H220.72888889c-24.27259221 0-48.54518557-24.27259221-48.54518556-48.54518556s24.27259221-48.54518557 48.54518556-48.54518444h582.54222222c24.27259221 0 48.54518557 24.27259221 48.54518556 48.54518444s-24.27259221 48.54518557-48.54518556 48.54518556z" fill="currentColor"></path>
</svg>
    <span>CATEGORYS</span>
  </div>
  <div class="widget-body">
    <ul class="categorys-list">
      
        <li class="categorys-list-item">
          <a href="/categories/web-%E5%BC%80%E5%8F%91/">
            web 开发 (3)
          </a>
        </li>
      
        <li class="categorys-list-item">
          <a href="/categories/%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90/">
            数据分析 (3)
          </a>
        </li>
      
        <li class="categorys-list-item">
          <a href="/categories/%E9%9A%8F%E7%AC%94/">
            随笔 (3)
          </a>
        </li>
      
        <li class="categorys-list-item">
          <a href="/categories/NLP/">
            NLP (1)
          </a>
        </li>
      
        <li class="categorys-list-item">
          <a href="/categories/%E7%AE%97%E6%B3%95/">
            算法 (1)
          </a>
        </li>
      
    </ul>
  </div>
</section>

      <section class="widget-tags widget-item  layout-margin content-padding--primary soft-size--large soft-style--box">
  <div class="widget-title">
    <svg class="icon icon-tags" viewBox="0 0 1098 1024" version="1.1" xmlns="http://www.w3.org/2000/svg">
  <path d="M283.42180005 272q0-28.38857157-20.09142843-48.48000001t-48.47999998-20.09142842-48.48000002 20.09142842-20.09142846 48.48000001 20.09142846 48.48 48.48000002 20.09142843 48.47999998-20.09142843 20.09142843-48.48zM855.0332285 580.57142843q0 28.38857157-19.81714313 48.2057147l-263.03999997 263.58857157q-20.9142853 19.81714313-48.75428534 19.81714312-28.38857157 0-48.20571468-19.81714312l-383.04-383.58857157q-20.36571468-19.81714313-34.55999999-54.10285688t-14.19428534-62.6742853l0-222.85714313q0-27.84000002 20.36571469-48.20571469t48.2057147-20.36571466l222.85714313 0q28.38857157 0 62.6742853 14.19428529t54.65142842 34.55999999l383.04000001 382.49142843q19.81714313 20.9142853 19.81714314 48.75428532zM1060.74751475 580.57142843q0 28.38857157-19.81714313 48.2057147l-263.04 263.58857157q-20.9142853 19.81714313-48.75428531 19.81714312-19.26857155 0-31.61142843-7.47428531t-28.38857159-24.13714314l251.79428534-251.7942853q19.81714313-19.81714313 19.81714308-48.20571469 0-27.84000002-19.81714308-48.75428531l-383.04000001-382.49142845q-20.36571468-20.36571468-54.65142842-34.55999999t-62.67428532-14.19428534l120 0q28.38857157 0 62.67428532 14.19428534t54.65142842 34.55999999l383.03999998 382.49142845q19.81714313 20.9142853 19.81714314 48.75428531z" fill="currentColor"></path>
</svg>
    <span>TAGS</span>
  </div>
  <div class="widget-body">
    <div class="tags-cloud">
      <a href="/tags/%E5%AE%8C%E6%88%90/" style="font-size: 20px;" class="tags-cloud-10">完成</a> <a href="/tags/%E8%BF%98%E6%B2%A1%E6%9C%89%E5%AE%8C%E6%88%90/" style="font-size: 10px;" class="tags-cloud-0">还没有完成</a>
    </div>
  </div>
</section>
    </div>
  </article>
</div>


    <!-- footer container -->
<footer id="footer" class="footer">
  <div class="footer-container">
    
    <div class="social-icons">
      
        
      
        
          <a href="https://www.zhihu.com/people/xing-zhe-80-14" class="soft-size--primary soft-style--box" target="_blank" rel="noopener noreferrer">
            <svg t="1587373160860" class="icon icon-zhihu" viewBox="0 0 1024 1024" version="1.1" xmlns="http://www.w3.org/2000/svg">
  <path d="M544.949 561.422s0-71.387-34.779-75.050c-34.779-3.663-142.775 0-142.775 0v-219.654h161.078s-1.83-73.219-32.949-73.219h-261.755l43.93-117.148s-65.897 3.663-89.692 45.761-98.844 252.604-98.844 252.604 25.627 10.983 67.726-20.134c42.101-31.116 56.743-86.033 56.743-86.033l76.879-3.663 1.83 223.316s-133.621-1.83-161.078 0c-27.457 1.83-42.101 75.050-42.101 75.050h203.182s-18.307 124.47-69.557 214.164c-53.085 89.692-151.929 161.078-151.929 161.078s71.387 29.287 140.947-10.983c69.557-42.101 120.811-223.316 120.811-223.316l162.912 203.182s14.643-97.013-1.83-124.47c-18.307-27.457-113.49-137.283-113.49-137.283l-42.101 36.607 29.287-120.811h177.552zM587.050 188.010l-1.83 660.793h65.897l23.795 82.37 115.321-82.37h162.912v-660.793h-366.091zM879.92 775.584h-76.879l-97.013 75.050-21.965-75.050h-20.134v-512.527h215.991v512.527z"></path>
</svg>
          </a>
        
      
        
      
        
          <a href="https://github.com/wenyuhaokikika" class="soft-size--primary soft-style--box" target="_blank" rel="noopener noreferrer">
            <svg class="icon icon-github" viewBox="0 0 1024 1024" version="1.1" xmlns="http://www.w3.org/2000/svg">
  <path d="M64.6 512c0 195.6 125.4 361.9 300.1 422.9 23.5 5.9 19.9-10.8 19.9-22.2v-77.6c-135.8 15.9-141.3-74-150.5-89-18.5-31.5-61.9-39.5-49-54.5 31-15.9 62.5 4 98.9 58 26.4 39.1 77.9 32.5 104.1 26 5.7-23.5 17.9-44.5 34.7-60.9-140.7-25.2-199.4-111.1-199.4-213.3 0-49.5 16.4-95.1 48.4-131.8-20.4-60.6 1.9-112.4 4.9-120.1 58.2-5.2 118.5 41.6 123.3 45.3 33.1-8.9 70.8-13.7 112.9-13.7 42.4 0 80.3 4.9 113.5 13.9 11.3-8.6 67.3-48.8 121.4-43.9 2.9 7.7 24.7 58.3 5.5 118.1 32.5 36.8 49 82.8 49 132.4 0 102.3-59 188.3-200.2 213.2 23.5 23.3 38.1 55.5 38.1 91.1v112.7c0.8 9 0 17.9 15.1 17.9C832.7 877 960.4 709.4 960.4 512.1c0-247.5-200.6-447.9-447.9-447.9C265 64.1 64.6 264.5 64.6 512z"></path>
</svg>
          </a>
        
      
    </div>
     
    <p>&copy; 2021 <a href="/" target="_blank">wenyuhao</a></p>

    

    <p>Powered by <a href="https://hexo.io" target="_blank" rel="noopener noreferrer">Hexo</a> Theme - <a href="https://github.com/miiiku/flex-block" target="_blank" rel="noopener noreferrer author">flex-block</a></p>

    <p>
      <a href="javascript:;" id="theme-light">🌞 浅色</a>
      <a href="javascript:;" id="theme-dark">🌛 深色</a>
      <a href="javascript:;" id="theme-auto">🤖️ 自动</a>
    </p>
  </div>
</footer>
  </div>

  <div class="back-to-top-fixed soft-size--round soft-style--box">
    <svg class="icon icon-back-to-top" viewBox="0 0 1024 1024" version="1.1" xmlns="http://www.w3.org/2000/svg">
      <path d="M725.333333 426.666667c-12.8 0-21.333333-4.266667-29.866667-12.8l-213.333333-213.333333c-17.066667-17.066667-17.066667-42.666667 0-59.733333s42.666667-17.066667 59.733333 0l213.333333 213.333333c17.066667 17.066667 17.066667 42.666667 0 59.733333C746.666667 422.4 738.133333 426.666667 725.333333 426.666667z"></path>
      <path d="M298.666667 426.666667c-12.8 0-21.333333-4.266667-29.866667-12.8-17.066667-17.066667-17.066667-42.666667 0-59.733333l213.333333-213.333333c17.066667-17.066667 42.666667-17.066667 59.733333 0s17.066667 42.666667 0 59.733333l-213.333333 213.333333C320 422.4 311.466667 426.666667 298.666667 426.666667z"></path>
      <path d="M512 896c-25.6 0-42.666667-17.066667-42.666667-42.666667L469.333333 170.666667c0-25.6 17.066667-42.666667 42.666667-42.666667s42.666667 17.066667 42.666667 42.666667l0 682.666667C554.666667 878.933333 537.6 896 512 896z"></path>
    </svg>
  </div>

  
  <!-- aplayer -->


<!-- dplayer -->




  


  


  




<script src="/js/script.js"></script>


  
  <!-- 尾部用户自定义相关内容 -->
</body>
</html>